app/usecase/keys/consumer.go

```diff
@@ -12,45 +12,69 @@ type Consumer interface {
 
 var _ Consumer = (*ConsumerPersist)(nil)
 
+type AvailableKeyRepoFactory func(tx transactional.Transaction) (repo.AvailableKey, error)
+type AllocatedKeyRepoFactory func(tx transactional.Transaction) (repo.AllocatedKey, error)
+
 type ConsumerPersist struct {
-	mutex            *sync.Mutex
-	availableKeyRepo repo.AvailableKey
-	allocatedKeyRepo repo.AllocatedKey
+	availableKeyRepoFactory AvailableKeyRepoFactory
+	allocatedKeyRepoFactory AllocatedKeyRepoFactory
+	transactionFactory      transactional.Factory
 }
 
-func (p ConsumerPersist) ConsumeInBatch(maxCount uint) ([]string, error) {
-	p.mutex.Lock()
-	defer p.mutex.Unlock()
+func (p ConsumerPersist) ConsumeInBatch(maxCount uint) (rawKeys []string, err error) {
+	tx, err := p.transactionFactory.Create()
 
-	keys, err := p.availableKeyRepo.RetrieveInBatch(maxCount)
 	if err != nil {
 		return nil, err
 	}
 
-	err = p.allocatedKeyRepo.CreateInBatch(keys)
-	if err != nil {
-		return nil, err
-	}
+	var keys []entity.Key
+
+	err = transactional.WithTransaction(tx, func(tx transactional.Transaction) error {
```

Comment:
It seems to me, that this approach is a little bit complicated. The interface
```go
type Transaction interface {
	Rollback() error
	Commit() error
}
```

tells how to work with transactions in an intuitive and simple manner. Anyway, we can easily change the interface if it will need.

Reply:
@alldroll I see. I'll rethink and do some experiment to see how can not avoid introducing unnecessary complexities while making it intuitive for users.

Reply:
@alldroll My intention was trying to have users only worry about `Transaction` or `transactional.WithTransaction`. Having to worry about multiple things together creates the needs for memorization ( IDE can auto complete this :) )